app/src/main/java/com/github/weg_li_android/MainActivity.kt

```diff
@@ -66,4 +265,163 @@ class MainActivity : AppCompatActivity() {
+        super.onActivityResult(requestCode, resultCode, data)
+        if (resultCode == Activity.RESULT_OK && data != null) {
+            when (requestCode) {
+                pickImage -> {
+                    if (data.clipData != null) {
+                        val count = data.clipData!!.itemCount
+                        for (i in 0 until count) {
+                            val imageUri: Uri? = data.clipData!!.getItemAt(i).uri
+                            val insertIndex = imageUri?.let { mainViewModel.addViolationPhoto(it) }
+                            if (insertIndex != null) {
+                                photoAdapter.notifyItemInserted(insertIndex)
+                            }
+                        }
+                    } else if (data.data != null) {
+                        val mImageUri: Uri? = data.data
+                        val insertIndex = mImageUri?.let { mainViewModel.addViolationPhoto(it) }
+                        if (insertIndex != null) {
+                            photoAdapter.notifyItemInserted(insertIndex)
+                        }
+                    }
+                }
+                takeImage -> {
+                    val bitmap = data.extras?.get("data") as Bitmap
+                    if (isStoragePermissionGranted()) {
+                        Timber.e("granted")
+                        saveAndInsertPhoto(bitmap)
+                    }
+                    else {
+                        tempBitmapOnRequestPermissions = bitmap
+                        Timber.e("Not granted")
+                    }
+                }
+            }
+        }
+    }
+
+    private fun saveAndInsertPhoto(bitmap : Bitmap) {
+        val imageUri : Uri = saveImageBitmap(bitmap, System.currentTimeMillis().toString())
+        Timber.e(imageUri.toString())
+        val insertIndex = mainViewModel.addViolationPhoto(imageUri)
+        photoAdapter.notifyItemInserted(insertIndex)
+    }
+
+    private fun isStoragePermissionGranted(): Boolean {
+        return if (Build.VERSION.SDK_INT >= 23) {
+            if (checkSelfPermission(Manifest.permission.WRITE_EXTERNAL_STORAGE)
+                == PackageManager.PERMISSION_GRANTED
+            ) {
+                Timber.v("Permission is granted")
+                true
+            } else {
+                Timber.v( "Permission is revoked")
+                ActivityCompat.requestPermissions(
+                    this,
+                    arrayOf(Manifest.permission.WRITE_EXTERNAL_STORAGE),
+                    permissionsRequestCode
+                )
+                (checkSelfPermission(Manifest.permission.WRITE_EXTERNAL_STORAGE)
+                        == PackageManager.PERMISSION_GRANTED)
+            }
+        } else { //permission is automatically granted on sdk<23 upon installation
+            Timber.v( "Permission is granted")
+            true
+        }
+    }
+
+    private fun saveImageBitmap(image_bitmap: Bitmap, image_name: String) : Uri {
```

Comment:
Use camel-case argument names on Kotlin and Java code :)  
Likely, it shouldn't be an Uri, but an Uri? because sometimes the code can fail.

Reply:
Also please have a look at kotlin's `use` method. That way, the `fos` is closed automatically.